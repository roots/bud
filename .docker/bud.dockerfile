FROM node:lts

LABEL name 'bud'
LABEL version 1

RUN npm install -g verdaccio pm2

COPY ./.docker/bud/motd /etc/motd
COPY ./.docker/bud/bash.bashrc /etc/bash.bashrc
