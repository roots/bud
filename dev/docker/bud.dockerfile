FROM node:16

LABEL name 'bud'
LABEL version 1

RUN npm install netlify-cli npm-cli-login --global

COPY ./dev/docker/bud/motd /etc/motd
COPY ./dev/docker/bud/bash.bashrc /etc/bash.bashrc

COPY ./ /srv/bud

RUN mkdir -p /srv/bud/node_modules \
  && mkdir -p /srv/bud/storage/verdaccio \
  && mkdir -p /srv/bud/storage/yarn \
  && mkdir -p /srv/bud/storage/coverage \
  && mkdir -p /srv/mocks
