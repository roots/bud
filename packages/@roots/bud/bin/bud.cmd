@echo off

node "%~dp0\bud.js" %*
