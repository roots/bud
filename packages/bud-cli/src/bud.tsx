#!/usr/bin/env node
import {yargs} from '@roots/bud-support'
import * as build from './commands/build'
import * as publish from './commands/publish'

/**
 * $bud
 */
const cli = yargs
  .command(build)
  .command(publish)
  .recommendCommands()
  .demandCommand(1, 'Try using one of the above commands.')
  .usage('bud [command] [options]')
  .version()
  .wrap(yargs.terminalWidth())
  .showHelpOnFail(true)
  .epilog('https://github.com/roots/bud').argv

export {cli as bud}
module.exports = {bud: cli}
