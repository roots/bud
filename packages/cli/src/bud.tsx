#!/usr/bin/env node

import {yargs} from '@roots/bud-support'
import * as build from './commands/build'
import * as publish from './commands/publish'

module.exports = yargs
  .command(build)
  .command(publish)
  .recommendCommands()
  .demandCommand(1, 'Try using one of the above commands.')
  .usage('bud [command] [options]')
  .version()
  .wrap(yargs.terminalWidth())
  .showHelpOnFail(true)
  .epilog('https://github.com/roots/bud').argv
