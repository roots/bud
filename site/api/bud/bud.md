---
id: bud.bud
title: bud namespace
sidebar_label: bud namespace
hide_title: true
sidebar: "api"
slug: bud
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

## Bud namespace

Signature:

```typescript
export declare namespace Bud
```

## Type Aliases

| Type Alias                      | Description             |
| ------------------------------- | ----------------------- |
| [Options](/api/bud/bud/options) | Bud constructor options |
