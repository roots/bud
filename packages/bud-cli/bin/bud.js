#!/usr/bin/node
require('./../lib/bud')
