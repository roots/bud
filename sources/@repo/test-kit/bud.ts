import * as CONSTANTS from '@repo/constants'
import {Bud, factory as budFactory, makeContext, seed} from '@roots/bud'
import {Config} from '@roots/bud-framework'
import {join} from 'node:path'

export const repoPath = (...path: Array<string>) =>
  join(CONSTANTS.REPO_PATH, ...(path ?? []))

export const mockProject = {
  path: repoPath('tests/util/project'),
}

export const factory = async (options?: Config.Options) => {
  const context = await makeContext(repoPath('tests/util/project'))

  const bud = await budFactory({
    name: 'bud',
    mode: 'production',
    ...(options ?? {}),
    context: {
      ...context,
      ...(options?.context ?? {}),
      args: {
        ...(options?.context?.args ?? {}),
        ci: true,
      },
    },
    seed: {
      ...seed,
      ...(options?.seed ?? {}),
    },
  })

  return bud
}

export {Bud}
