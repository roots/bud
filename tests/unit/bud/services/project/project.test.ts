import {fs} from '@roots/bud-support'
import {Bud, factory} from '@repo/test-kit/bud'
import {expect, describe, beforeAll, it} from '@jest/globals'

const PROJECT_MANIFEST_PATH = `${process.cwd()}/tests/util/project/package.json`
const PROJECT_BUD_PROFILE_PATH = `${process.cwd()}/tests/util/project/.budfiles/bud/profile.json`

describe('bud.project', function () {
  let bud: Bud
  let manifest: Record<string, any>
  let profile: Record<string, any>

  beforeAll(async () => {
    bud = await factory()
    await bud.build.make()
    manifest = await fs.readJson(PROJECT_MANIFEST_PATH)
    profile = await fs.readJson(PROJECT_BUD_PROFILE_PATH)
  })

  it('holds cache validation hash', () => {
    expect(bud.project.get('cache')).toMatchSnapshot({
      hash: expect.any(String),
    })
  })

  it('holds cache deps', async () => {
    expect(bud.project.get('dependencies').length).toEqual(2)
  })

  it('holds env values', async () => {
    expect(bud.project.get('env.all')).toMatchSnapshot({
      TEST: 'VALUE',
      PUBLIC_APP_TEST: 'PUBLIC_VALUE',
    })

    expect(bud.project.get('env.public')).toMatchSnapshot({
      APP_TEST: 'PUBLIC_VALUE',
    })
  })

  it('has evn records matching profile.json artifact', async () => {
    expect(bud.project.get('env.all')).toMatchSnapshot(
      profile.env.all,
    )
    expect(bud.project.get('env.public')).toMatchSnapshot(
      profile.env.public,
    )
  })

  it('holds cli records', async () => {
    expect(bud.project.get('cli')).toStrictEqual({
      args: {},
      argv: [],
      flags: {},
      metadata: {},
      raw: [],
    })
  })

  it('holds manifest records', async () => {
    expect(bud.project.get('manifest')).toStrictEqual(manifest)
  })

  it('references @roots/bud-babel', async () => {
    expect(
      bud.project.get('modules.@roots/bud-babel'),
    ).toMatchSnapshot({
      bud: {type: 'extension'},
      name: '@roots/bud-babel',
      peerDependencies: {},
      requires: [],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references @roots/bud-entrypoints', async () => {
    expect(
      bud.project.get('modules.@roots/bud-entrypoints'),
    ).toMatchSnapshot({
      bud: {
        type: 'extension',
      },
      name: '@roots/bud-entrypoints',
      peerDependencies: {},
      requires: [],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references @roots/bud-eslint', async () => {
    expect(
      bud.project.get('modules.@roots/bud-eslint'),
    ).toMatchSnapshot({
      bud: {
        type: 'extension',
      },
      name: '@roots/bud-eslint',
      peerDependencies: {
        eslint: expect.any(String),
      },
      requires: [],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references @roots/bud-postcss', async () => {
    expect(
      bud.project.get('modules.@roots/bud-postcss'),
    ).toMatchSnapshot({
      bud: {
        type: 'extension',
      },
      name: '@roots/bud-postcss',
      peerDependencies: {
        postcss: expect.any(String),
        'postcss-import': expect.any(String),
        'postcss-nested': expect.any(String),
        'postcss-preset-env': expect.any(String),
      },
      requires: [],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references @roots/bud-entrypoints', async () => {
    expect(
      bud.project.get('modules.@roots/bud-entrypoints'),
    ).toMatchSnapshot({
      bud: {
        type: 'extension',
      },
      name: '@roots/bud-entrypoints',
      peerDependencies: {},
      requires: [],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references @roots/bud-preset-recommend', async () => {
    expect(
      bud.project.get('modules.@roots/bud-preset-recommend'),
    ).toMatchSnapshot({
      bud: {
        peers: [
          '@roots/bud-babel',
          '@roots/bud-entrypoints',
          '@roots/bud-postcss',
        ],
        type: 'extension',
      },
      name: '@roots/bud-preset-recommend',
      peerDependencies: {},
      requires: [
        ['@roots/bud-babel', expect.any(String)],
        ['@roots/bud-entrypoints', expect.any(String)],
        ['@roots/bud-postcss', expect.any(String)],
      ],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references the root package', async () => {
    expect(bud.project.get('modules.root')).toMatchSnapshot({
      browserslist: ['extends @wordpress/browserslist-config'],
      bud: null,
      devDependencies: {
        '@roots/bud': 'workspace:*',
        '@roots/bud-babel': 'workspace:*',
        '@roots/bud-eslint': 'workspace:*',
        '@roots/bud-postcss': 'workspace:*',
        '@roots/bud-preset-recommend': 'workspace:*',
        '@roots/bud-tailwindcss': 'workspace:*',
      },
      name: 'root',
      parent: null,
      private: true,
      requires: [
        ['@roots/bud', 'workspace:*'],
        ['@roots/bud-babel', 'workspace:*'],
        ['@roots/bud-eslint', 'workspace:*'],
        ['@roots/bud-postcss', 'workspace:*'],
        ['@roots/bud-preset-recommend', 'workspace:*'],
        ['@roots/bud-tailwindcss', 'workspace:*'],
      ],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('references @roots/bud-tailwindcss', async () => {
    expect(
      bud.project.get('modules.@roots/bud-tailwindcss'),
    ).toMatchSnapshot({
      bud: {
        peers: ['@roots/bud-postcss'],
        type: 'extension',
      },
      name: '@roots/bud-tailwindcss',
      peerDependencies: {
        tailwindcss: expect.any(String),
      },
      requires: [['@roots/bud-postcss', expect.any(String)]],
      resolvable: true,
      version: expect.any(String),
    })
  })

  it('resolveFrom contains paths of found peers', async () => {
    expect(bud.build.config.resolve.modules).toMatchSnapshot([
      expect.stringContaining('src'),
      expect.stringContaining('node_modules'),
      expect.stringContaining('/sources'),
    ])
  })
})
