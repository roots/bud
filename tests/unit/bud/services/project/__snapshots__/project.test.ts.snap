// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bud.project has evn records matching profile.json artifact 1`] = `
Object {
  "PUBLIC_APP_TEST": "PUBLIC_VALUE",
  "TEST": "VALUE",
}
`;

exports[`bud.project has evn records matching profile.json artifact 2`] = `
Object {
  "APP_TEST": "PUBLIC_VALUE",
}
`;

exports[`bud.project holds cache validation hash 1`] = `
Object {
  "hash": Any<String>,
}
`;

exports[`bud.project holds env values 1`] = `
Object {
  "PUBLIC_APP_TEST": "PUBLIC_VALUE",
  "TEST": "VALUE",
}
`;

exports[`bud.project holds env values 2`] = `
Object {
  "APP_TEST": "PUBLIC_VALUE",
}
`;

exports[`bud.project references @roots/bud-babel 1`] = `
Object {
  "bud": Object {
    "type": "extension",
  },
  "name": "@roots/bud-babel",
  "peerDependencies": Object {
    "@babel/core": "^7.16.10",
    "@babel/plugin-proposal-class-properties": "^7.16.0",
    "@babel/plugin-proposal-object-rest-spread": "^7.16.0",
    "@babel/plugin-syntax-dynamic-import": "^7.8.3",
    "@babel/plugin-transform-runtime": "^7.16.4",
    "@babel/preset-env": "^7.16.4",
    "babel-loader": "^8.2.3",
    "babel-plugin-add-module-exports": "^1.0.4",
  },
  "requires": Array [],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references @roots/bud-entrypoints 1`] = `
Object {
  "bud": Object {
    "type": "extension",
  },
  "name": "@roots/bud-entrypoints",
  "peerDependencies": Object {},
  "requires": Array [],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references @roots/bud-entrypoints 2`] = `
Object {
  "bud": Object {
    "type": "extension",
  },
  "name": "@roots/bud-entrypoints",
  "peerDependencies": Object {},
  "requires": Array [],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references @roots/bud-eslint 1`] = `
Object {
  "bud": Object {
    "type": "extension",
  },
  "name": "@roots/bud-eslint",
  "peerDependencies": Object {
    "eslint": Any<String>,
  },
  "requires": Array [],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references @roots/bud-postcss 1`] = `
Object {
  "bud": Object {
    "type": "extension",
  },
  "name": "@roots/bud-postcss",
  "peerDependencies": Object {
    "postcss": Any<String>,
    "postcss-import": Any<String>,
    "postcss-nested": Any<String>,
    "postcss-preset-env": Any<String>,
  },
  "requires": Array [],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references @roots/bud-preset-recommend 1`] = `
Object {
  "bud": Object {
    "peers": Array [
      "@roots/bud-babel",
      "@roots/bud-entrypoints",
      "@roots/bud-postcss",
    ],
    "type": "extension",
  },
  "name": "@roots/bud-preset-recommend",
  "peerDependencies": Object {},
  "requires": Array [
    Array [
      "@roots/bud-babel",
      Any<String>,
    ],
    Array [
      "@roots/bud-entrypoints",
      Any<String>,
    ],
    Array [
      "@roots/bud-postcss",
      Any<String>,
    ],
  ],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references @roots/bud-tailwindcss 1`] = `
Object {
  "bud": Object {
    "peers": Array [
      "@roots/bud-postcss",
    ],
    "type": "extension",
  },
  "name": "@roots/bud-tailwindcss",
  "peerDependencies": Object {
    "tailwindcss": Any<String>,
  },
  "requires": Array [
    Array [
      "@roots/bud-postcss",
      Any<String>,
    ],
  ],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project references the root package 1`] = `
Object {
  "browserslist": Array [
    "extends @wordpress/browserslist-config",
  ],
  "bud": null,
  "devDependencies": Object {
    "@roots/bud": "workspace:*",
    "@roots/bud-babel": "workspace:*",
    "@roots/bud-eslint": "workspace:*",
    "@roots/bud-postcss": "workspace:*",
    "@roots/bud-preset-recommend": "workspace:*",
    "@roots/bud-tailwindcss": "workspace:*",
  },
  "name": "root",
  "parent": null,
  "private": true,
  "requires": Array [
    Array [
      "@roots/bud",
      "workspace:*",
    ],
    Array [
      "@roots/bud-babel",
      "workspace:*",
    ],
    Array [
      "@roots/bud-eslint",
      "workspace:*",
    ],
    Array [
      "@roots/bud-postcss",
      "workspace:*",
    ],
    Array [
      "@roots/bud-preset-recommend",
      "workspace:*",
    ],
    Array [
      "@roots/bud-tailwindcss",
      "workspace:*",
    ],
  ],
  "resolvable": true,
  "version": Any<String>,
}
`;

exports[`bud.project resolveFrom contains paths of found peers 1`] = `
Array [
  StringContaining "src",
  StringContaining "node_modules",
  StringContaining "/sources",
]
`;
